## Requirements

Physcraper requires the user to install:

- [Anaconda](https://docs.anaconda.com/anaconda/install/): *Anaconda Software Distribution. Computer software. Vers. 4.8.0. Anaconda, July. 2021. Web. <https://anaconda.com>*
- OR [Virtualenv](https://pypi.org/project/virtualenv/),
- [MUSCLE](https://www.ebi.ac.uk/Tools/msa/muscle/): *Edgar RC. MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Res. 2004 Mar 19;32(5):1792-7. doi: [10.1093/nar/gkh340](https://doi.org/10.1093/nar/gkh340)*
- AND [RAxML](https://cme.h-its.org/exelixis/web/software/raxml/): *Stamatakis, Alexandros. "RAxML version 8: a tool for phylogenetic analysis and post-analysis of large phylogenies." Bioinformatics 30.9 (2014): 1312-1313.* doi: [10.1093/bioinformatics/btu033](https://doi.org/10.1093/bioinformatics/btu033)


Physcraper relies on the following Python packages that are automatically installed:

- argparse
- biopython: *Cock, P. J., Antao, T., Chang, J. T., Chapman, B. A., Cox, C. J., Dalke, A., … others. (2009). Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics, 25(11), 1422–1423.*
- configparser
- coverage
- DateTime
- [DendroPy](https://dendropy.org/primer/index.html): *Sukumaran, J and MT Holder. 2010. DendroPy: a Python library for phylogenetic computing. Bioinformatics 26: 1569-1571* doi:  [10.1093/bioinformatics/btq228](https://doi.org/10.1093/bioinformatics/btq228)
- future
- m2r2
- nexson
- numpy: *Harris, C. R., Millman, K. J., van der Walt, S. J., Gommers, R., Virtanen, P., Cournapeau, D., … Oliphant, T. E. (2020). Array programming with NumPy. Nature, 585, 357–362.* doi: [10.1038/s41586-020-2649-2](https://doi.org/10.1038/s41586-020-2649-2)
- [OpenTree](https://github.com/OpenTreeOfLife/python-opentree): *Emily Jane McTavish, Luna Luisa Sanchez-Reyes, Mark T. Holder. (2020). OpenTree: A Python package for Accessing and Analyzing data from the Open Tree of Life. BioRxiv 2020.12.14.422759* doi: [10.1101/2020.12.14.422759](https://doi.org/10.1101/2020.12.14.422759)
- pandas: *McKinney, W., & others. (2010). Data structures for statistical computing in python. In Proceedings of the 9th Python in Science Conference (Vol. 445, pp. 51–56).*
- pytest
- pytest-cov
- pytest-xdist
- recommonmark
- requests: *Chandra, R. V., & Varanasi, B. S. (2015). Python requests essentials. Packt Publishing Ltd.*
- sh
- sphinx
- urllib3
